### PR TITLE
fixes #1089, #1090

### LINKED ISSUE #1089: Out of bounds access in SteadystateProblem::workSteadyStateProblem if NewtonMaxSteps is 0
Integration errors during preequilibration lead to out of bounds access in case NewtonMaxSteps is set to 0.

In `steadystateproblem.cpp`:
* if `solver.getNewtonMaxSteps()` is zero
* `SteadystateProblem::numlinsteps` will be of size zero
* https://github.com/ICB-DCM/AMICI/blob/5af8725d98ed90b8ebb3af68a7e4b95386e8b78c/src/steadystateproblem.cpp#L90-L98
  will access `numlinsteps[1]` -> invalid

### PR DESCRIPTION
**Pull request by paulstapor**
in newtonsolver and steadystateproblem

### CODE REVIEW

File: src/steadystateproblem.cpp [hunk 1]
```diff
@@ -24,7 +24,7 @@ SteadystateProblem::SteadystateProblem(const Solver &solver, const Model &model)
       xdot(model.nx_solver), xdot_old(model.nx_solver),
       sx(model.nx_solver, model.nplist()), sdx(model.nx_solver, model.nplist()),
       dJydx(model.nJ * model.nx_solver * model.nt(), 0.0), numsteps(3, 0),
-      numlinsteps(solver.getNewtonMaxSteps() * 2, 0) {}
+      numlinsteps(std::max(solver.getNewtonMaxSteps(), 1) * 2, 0) {}
```
Comment (dweindl) [inline src/steadystateproblem.cpp:27]: Why not have maxsteps as member of SteadystateProblem and enforcing >=1 there instead of using max in 5 places?

Comment (paulstapor) [reply src/steadystateproblem.cpp]: Yes, looks nicer...
Thanks!

File: src/newton_solver.cpp [hunk 2]
```diff
@@ -75,7 +75,11 @@ std::unique_ptr<NewtonSolver> NewtonSolver::getSolver(realtype *t, AmiVector *x,
     solver->dampingFactorMode = simulationSolver.getNewtonDampingFactorMode();
     solver->dampingFactorLowerBound =
         simulationSolver.getNewtonDampingFactorLowerBound();
-    solver->numlinsteps.resize(simulationSolver.getNewtonMaxSteps(), 0.0);
+    if (simulationSolver.getLinearSolver() == LinearSolver::SPBCG) {
+        solver->numlinsteps.resize(simulationSolver.getNewtonMaxSteps(), 0.0);
```
Comment (dweindl) [inline src/newton_solver.cpp:79]:
```suggestion
        solver->numlinsteps.resize(simulationSolver.getNewtonMaxSteps(), 0);
```

Was already there before, but as this is an `int` array, using double seems unnecessarily confusing...

File: src/newton_solver.cpp [hunk 3]
```diff
@@ -75,7 +75,11 @@ std::unique_ptr<NewtonSolver> NewtonSolver::getSolver(realtype *t, AmiVector *x,
...
+    } else {
+        solver->numlinsteps.resize(1, 0.0);
```
Comment (dweindl) [inline src/newton_solver.cpp:81]:
```suggestion
        solver->numlinsteps.resize(1, 0);
```

File: src/steadystateproblem.cpp [hunk 4]
```diff
@@ -23,8 +23,13 @@ SteadystateProblem::SteadystateProblem(const Solver &solver, const Model &model)
       x(model.nx_solver), x_old(model.nx_solver), dx(model.nx_solver),
       xdot(model.nx_solver), xdot_old(model.nx_solver),
       sx(model.nx_solver, model.nplist()), sdx(model.nx_solver, model.nplist()),
+      maxSteps(std::max(solver.getNewtonMaxSteps(), 1)),
       dJydx(model.nJ * model.nx_solver * model.nt(), 0.0), numsteps(3, 0),
-      numlinsteps(solver.getNewtonMaxSteps() * 2, 0) {}
+      numlinsteps(2, 0) {
+          /* maxSteps must be adapted if iterative linear solvers are used */
+          if (solver.getLinearSolver() == LinearSolver::SPBCG)
+              numlinsteps.resize(2 * solver.getNewtonMaxSteps(), 0.0);
```
Comment (dweindl) [inline src/steadystateproblem.cpp:31]:
```suggestion
              numlinsteps.resize(2 * solver.getNewtonMaxSteps(), 0);
```

File: src/newton_solver.cpp (hunk 2 above)
Comment (paulstapor) [reply src/newton_solver.cpp:79]: ah, true, sorry...

**Comment by paulstapor** [comment]
What about the failing C++ testsuite? Is there currently some issues linked to #1095 ?

**Comment by dweindl** [comment]
> What about the failing C++ testsuite? Is there currently some issues linked to #1095 ?

I don't think so. Should all be related to changes here.

File: src/steadystateproblem.cpp [hunk 5]
```diff
@@ -23,8 +23,13 @@ SteadystateProblem::SteadystateProblem(const Solver &solver, const Model &model)
       x(model.nx_solver), x_old(model.nx_solver), dx(model.nx_solver),
       xdot(model.nx_solver), xdot_old(model.nx_solver),
       sx(model.nx_solver, model.nplist()), sdx(model.nx_solver, model.nplist()),
+      maxSteps(std::max(solver.getNewtonMaxSteps(), 1)),
       dJydx(model.nJ * model.nx_solver * model.nt(), 0.0), numsteps(3, 0),
-      numlinsteps(solver.getNewtonMaxSteps() * 2, 0) {}
+      numlinsteps(2, 0) {
+          /* maxSteps must be adapted if iterative linear solvers are used */
+          if (solver.getLinearSolver() == LinearSolver::SPBCG)
+              numlinsteps.resize(2 * solver.getNewtonMaxSteps(), 0);
```
Comment (dweindl) [inline src/steadystateproblem.cpp:31]: This doesn't handle the 0 in solver.getNewtonMaxSteps

File: src/steadystateproblem.cpp [hunk 6]
```diff
@@ -88,13 +93,11 @@ void SteadystateProblem::workSteadyStateProblem(Solver *solver, Model *model,
...
-                    newtonSolver->getNumLinSteps().begin(),
-                    solver->getNewtonMaxSteps(),
+                    newtonSolver->getNumLinSteps().begin(), maxSteps,
                     &numlinsteps.at(solver->getNewtonMaxLinearSteps() + 1));
             } catch (NewtonFailure const &ex3) {
                 std::copy_n(
-                    newtonSolver->getNumLinSteps().begin(),
-                    solver->getNewtonMaxSteps(),
+                    newtonSolver->getNumLinSteps().begin(), maxSteps,
                     &numlinsteps.at(solver->getNewtonMaxLinearSteps() + 1));
```
Comment (dweindl) [inline src/steadystateproblem.cpp:101]: Why the +1 here? Isn't that out of bounds?

File: src/steadystateproblem.cpp (hunk 5 above)
Comment (paulstapor) [reply src/steadystateproblem.cpp]: Technically it's valid, although only possible when misspecified by the user... 
Yet: Cleaner to catch it here, I think...

**Comment by paulstapor** [comment]
Okay, sonarcloud code smell will be soon addressed anyway, major stuff is addressed however. Waiting for travis. Would merge then, once travis has passed.
Fine for everyone, also that sympy is pinned to 1.5.1?

**Comment by dweindl** [comment]
> Would merge then

Okay for me. As @FFroehlich meanwhile fixed #1095 in #1098, sympy needs to be unpinned there again.

**Pull request merged**